doc/design.md

```diff
@@ -0,0 +1,178 @@
+---
+title: "Design"
+date: 2023-09-01
+description: "Overall API & Pipeline design"
+weight: 1
+---
+
+## API architecture
+
+The main building blocks of the API are put together this way:
+
+```mermaid
+graph TD
+  api(FastAPI)
+  pubsub(Pub/Sub) --- api
+  redis(Redis) --- pubsub
+  models(Pydantic) --- api
+  db(MongoDB) --- models
+
+  storage(Storage)
+
+  api --> client
+  storage --> client
+```
+
+The `client` component can be a number of things.  Each pipeline step is a
+client, and there could be extra standalone ones too.  They interact via the
+API and the storage services separately.  By default, storage is provided by
+SSH for uploads and HTTP for downloads for standalone deployments using
+`docker-compose`.  Production deployments may use alternative solutions as
+long as all the binaries are accessible via direct HTTP URLs.
+
+## Why a new API?
+
+The new KernelCI API is a complete redesign to replace the legacy
+[backend](https://api.kernelci.org/) used in production, which has several
+limitations.  In particular, the [**previous backend
+API**](https://github.com/kernelci/kernelci-backend):
+
+* was written in Python 2.7, which has now reached end-of-life
+* has a monolithic design with many built-in features (regression tracking,
+  email reports, parsing test results directly from LAVA labs...)
+* uses Celery for asynchronous request handling, whereas modern Python can do
+  this natively
+* has no pub/sub mechanism, orchestration relies on an external framework
+  (i.e. Jenkins)
+
+To overcome these limitations, the **new API** has the following
+characteristics:
+
+* based on [FastAPI](https://fastapi.tiangolo.com/) to provide native
+  asynchronous request handling, data model validation using
+  [Pydantic](https://pydantic-docs.helpmanual.io/), automatically generated
+  documentation with [OpenAPI](https://www.openapis.org/).  See also the
+  [OpenAPI JSON
+  description](https://staging.kernelci.org:9000/latest/openapi.json).
+* Pub/Sub mechanism via the API, with [Redis](https://redis.io/) to manage
+  message queues.  This can now be used to coordinate client-side functions and
+  recreate a full modular pipeline with no additional framework
+* [CloudEvent](https://cloudevents.io/) for the formatting of Pub/Sub events
+* is written for [Python 3.10](https://www.python.org/downloads/release/python-3100/)
+* relies on [JWT](https://jwt.io/) and [OAuth 2.0](https://oauth.net/2/)
+  authentication for inter-operability with other services
+* treats storage entirely separately from the API which purely handles data.
+  Several storage solutions can be supported, such as SSH for self-contained
+  systems, AzureFiles and S3 for Cloud deployments etc.
```

Comment:
```suggestion
  systems, AzureFiles, and S3 for Cloud deployments etc.
```

Reply:
Well this one is a matter of style, see [Oxford comma](https://en.wikipedia.org/wiki/Serial_comma).  But yes I would rather drop it in this case.